### PR TITLE
Make `slice()` torch-scriptable

### PR DESCRIPTION
**Pull request by frostedoyster**
Failing at the moment due to a weird error given by `check_isinstance`.

### CODE REVIEW

File: python/metatensor-operations/metatensor/operations/_dispatch.py [hunk 1]
```diff
@@ -113,6 +113,19 @@ def bincount(input, weights: Optional[TorchTensor] = None, minlength: int = 0):
...
 
+def bool_array_like(bool_list: List[bool], like):
+    """ 
+    Converts the input list of bool to a numpy array or torch tensor
+    based on the type of `like`.
+    """
+    if isinstance(like, TorchTensor):
+        return torch.tensor(bool_list, device=like.device)
+    elif isinstance(like, np.ndarray):
+        return np.array(bool_list)
```
Comment (Luthaf) [inline python/metatensor-operations/metatensor/operations/_dispatch.py:124]: should this set a `dtype=np.bool/torch.bool` as well? Just to catch mistakes early if needed

File: python/metatensor-operations/metatensor/operations/_dispatch.py [hunk 2]
```diff
@@ -136,6 +149,19 @@ def eye_like(array, size: int):
...
+    """
+    if isinstance(like, TorchTensor):
+        return torch.full(shape, fill_value, dtype=like.dtype, device=like.device)
+    elif isinstance(like, np.ndarray):
+        return np.full(shape, fill_value)
+    else:
+        raise TypeError(UNKNOWN_ARRAY_TYPE)
 
 
 def list_to_array(array, data: Union[List[int], List[List[int]]]):
```
Comment (Luthaf) [inline python/metatensor-operations/metatensor/operations/_dispatch.py:167]: should we rename this one to `int_array_like`?

File: python/metatensor-operations/metatensor/operations/_dispatch.py [hunk 3]
```diff
@@ -146,6 +172,23 @@ def list_to_array(array, data: Union[List[int], List[List[int]]]):
         return np.array(data)
     else:
         raise TypeError(UNKNOWN_ARRAY_TYPE)
+    
+
+def mask(array, axis: int, mask):
```
Comment (Luthaf) [inline python/metatensor-operations/metatensor/operations/_dispatch.py:177]: what's the issue with `array[mask]`?

File: python/metatensor-operations/tests/split.py [hunk 4]
```diff
@@ -520,7 +516,7 @@ def test_split_errors(self):
             ),
         self.assertEqual(
             str(cm.exception),
-            "each element in grouped_labels must be 'Labels', not <class 'str'>",
```
Comment (Luthaf) [inline python/metatensor-operations/tests/split.py:523]: Instead of removing the error message context, the type check should only be done if not in torch script mode (since in torch script mode the types are checked statically)

File: python/metatensor-operations/metatensor/operations/_dispatch.py (hunk 1 above)
Comment (frostedoyster) [reply python/metatensor-operations/metatensor/operations/_dispatch.py:124]: Good idea

File: python/metatensor-operations/metatensor/operations/_dispatch.py (hunk 2 above)
Comment (frostedoyster) [reply python/metatensor-operations/metatensor/operations/_dispatch.py:167]: But it can also be a float, no?

Comment (frostedoyster) [reply python/metatensor-operations/metatensor/operations/_dispatch.py:167]: But it can also be a float, no?

File: python/metatensor-operations/metatensor/operations/_dispatch.py (hunk 3 above)
Comment (frostedoyster) [reply python/metatensor-operations/metatensor/operations/_dispatch.py:177]: The issue is when doing properties. In that case, you would really like to do `array[..., mask]`, but indexing after an ellipsis is not supported in torchscript

File: python/metatensor-operations/metatensor/operations/_dispatch.py (hunk 2 above)
Comment (Luthaf) [reply python/metatensor-operations/metatensor/operations/_dispatch.py:167]: data is defined a `int`, and I don't think we ever call it with float

File: python/metatensor-operations/metatensor/operations/_dispatch.py (hunk 3 above)
Comment (Luthaf) [reply python/metatensor-operations/metatensor/operations/_dispatch.py:177]: ok, fair enough. Might be worth adding a small comment explaining that to the next person who will touch this code!

File: python/metatensor-operations/metatensor/operations/_dispatch.py (hunk 2 above)
Comment (frostedoyster) [reply python/metatensor-operations/metatensor/operations/_dispatch.py:167]: Ok

File: python/metatensor-operations/metatensor/operations/_dispatch.py (hunk 3 above)
Comment (frostedoyster) [reply python/metatensor-operations/metatensor/operations/_dispatch.py:177]: Will add the comment

**Comment by Luthaf** [comment]
Rebased & force pushed

**Review by Luthaf — APPROVED** [review]
